### ISSUE TITLE
Replace dict.has_key() with the 'in' keyword, for twisted/mail

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8361]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8361|
|Type|enhancement|
|Created|2016-05-26 17:37:36Z|
|Branch|https://github.com/twisted/twisted/tree/8361-rodrigc-has_key3|

In PEP 0290, has_key() was deprecated in Python 2.2 and higher: 
* https://www.python.org/dev/peps/pep-0290/#testing-dictionary-membership
* https://docs.python.org/2.2/whatsnew/node4.html

In Python 3, dict.has_key() was removed:
* https://docs.python.org/3.0/whatsnew/3.0.html#builtins

PEP 0290 says that for testing dictionary membership, this:

```
if d.has_key(k):
```

should be changed to this:

```
if k in d:
```

<details><summary>Searchable metadata</summary>

```
trac-id__8361 8361
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8361_rodrigc_has_key3 8361-rodrigc-has_key3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__None None
time__1464284256439998 1464284256439998
changetime__1465639694271234 1465639694271234
version__None None
owner__Amber_Brown__HawkOwl___hawkowl_____ Amber Brown (HawkOwl) <hawkowl@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/77

This was pulled out from [ticket:4053]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

While reviewing this I saw twisted.mail.mail.DomainWithDefaultDict.has_key which always returns 1, but is documented to return a dict... I think that it should be deprecated.

since this patch is small I think that we can use the same ticket for also deprecating it.

I have pushed a commit which deprecates the has_key.

Sending to buildbot and leaving this for review as I have also contributed to this ticket and therefore we need another developer to check the patch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to @hawkowl<br>@hawkowl set status to `closed`|
|-|-|

In changeset 60de1ebda4cdf28e028d9342c5769475046af9b7
```
#!CommitTicketReference repository="" revision="60de1ebda4cdf28e028d9342c5769475046af9b7"
Merge 8361-rodrigc-has_key3: Remove has_key() from twisted/mail

Author: rodrigc, adiroiban
Reviewer: hawkowl
Fixes: #8361
```
